### DISCUSSION TITLE
IPv6 + DNS records

### DISCUSSION THREAD

**Discussion by passing** [jamulussoftware/jamulus#2170] [General]
Hi,

I have setup a Jamulus Server (at AWS in Frankfurt) with IPv6 support and configured 3 DNS records:
- public.jamulus.click - IPv4+IPv6
- public4.jamulus.click - IPv4 only
- public6.jamulus.click - IPv6 only

Now while testing, I heard from some users of ISP 'Deutsche Telekom AG', that the first DNS record - which resolves both IPv4 and IPv6 addresses - doesn't work for them at all, while the IPv4-only one works.

In general it would be nice if you could have just one DNS record that works for everybody. Is anybody able to reconstruct that issue and maybe has an idea what's wrong with using a combined IPv4/IPv6 DNS record?

Also, I'd be curious, if anybody with a DSlite connection can test, how much the delay is affected when connecting via IPv6.

Cheers,
Martin

**Comment by rdica** [comment]
I was able to connect fine via IPv6 to `public.jamulus.click`, I have native IPv6 from my ISP in US. I use Linux.

You should get more information from those users, "doesn't work for them at all" provides no details with which to troubleshoot. My thought on that is "your ISP has a broken IPv6 implementation, and/or bad firewalling, and/or you have an old OS, did you start jamulus with -6"....just too many things to guess at.

One other question comes to mind, does jamulus code implement Happy Eyeballs? https://en.wikipedia.org/wiki/Happy_Eyeballs

I should also note I too run a couple servers in US that are IPv6 capable, and use a single hostname for A and AAAA records. It's up to the client and its OS to determine which protocol is used.

**Reply by pljones** [reply]
No, a Jamulus instance (client or server) is only either IPv4 or IPv6, never both.

As you note, DNS can provide either address for a single hostname and you could run two server instances on separate port numbers on a single machine, one being IPv4, one IPv6.  But - currently - there's no automatic switching.

Further IPv6 support is hoped for - keeping the protocol backward-compatible in both directions (i.e. new client<->old server and old client<->new server, plus similar complexity around the directories) means it's not trivial, of course.

**Reply by rdica** [reply]
Not sure what you're saying is correct.
I dont need to run separate server instances based on IP protocol.
A single server started with `-6` listens and accepts connections for both v4 and v6 connections.
My client, invoked with `-6` connects to v4 and v6 servers.

I've never tested happy eyeballs connectivity with the client hence my question regarding it.
The switching I think you're referring to is actually hostname based happy eyeballs as noted in the wikipedia link I posted above.

**Reply by pljones** [reply]
I didn't check the source - I was doing it from memory.  I should have read the help output, at least -- "IPv4 is always enabled"... :)

**Reply by passing** [reply]
> You should get more information from those users, "doesn't work for them at all" provides no details with which to troubleshoot. My thought on that is "your ISP has a broken IPv6 implementation, and/or bad firewalling, and/or you have an old OS, did you start jamulus with -6"....just too many things to guess at.

"doesn't work for them at all" means they did enter the server address and trigger the 'Connect' Button, which brought them back to the main window without being connected (which is what you get when you enter an invalid address as reported in https://github.com/jamulussoftware/jamulus/issues/1945)
And they did not run Jamulus in IPv6 mode.

I thought that there might be something wrong with the combined IPv4/IPv6 DNS record, but in fact Jamulus running without `-6` should not care about that at all. Anyways, I asked them to check if the issue is still there..

**Reply by rdica** [reply]
As part of my initial testing when IPv6 was first added in the 3.8.1beta, I tested using a FQDN as server address, with and without `-6` and got the expected behaviour; with `-6` the client used the IPv6 IP, without it the v4 IP was used.

If you want to test, use server address `daw4.diix.org`.

**Comment by ghost** [comment]
@rdica asked: One other question comes to mind, does jamulus code implement Happy Eyeballs?
The operating system has the final say here. Jamulus code will have to be aware of system config files, DNS caches, etc. Best to keep it simple.

**Comment by ghost** [comment]
Here is a protocol for Happy Ear Lobes:
Jamulus responds to all ip-addresses returned by all DNS A and AAAA records, then uses the One ipv4/6 address with the lowest Ping.
